File: LinkedList/82_RemoveDuplicatesFromSortedListII.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com
# @Last Modified time: 2016-04-29 17:15:38


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


# Recursively
class Solution(object):
    def deleteDuplicates(self, head):
        if not head or not head.next:
            return head
        if head.val == head.next.val:
            while head.next and head.val == head.next.val:
                head = head.next
            return self.deleteDuplicates(head.next)
        else:
            head.next = self.deleteDuplicates(head.next)
            return head


# Iteraively
class Solution_2(object):
    def deleteDuplicates(self, head):
        cur = pre_head = ListNode(0)
        while head:
            if head.next and head.val == head.next.val:
                # Skip the duplicated nodes.
                while head.next and head.val == head.next.val:
                    head = head.next
                head = head.next
            # we can make sure head is one single node here.
            else:
                cur.next = head
                cur = cur.next
                head = head.next
        cur.next = None     # Make sure the cur here is the tail: [1,2,2]
        return pre_head.next

"""
[]
[1]
[1,2,2]
[3,3,3,3,3]
[1,1,1,2,3,4,4,4,4,5]
"""

File: LinkedList/82_RemoveDuplicatesFromSortedListII.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-29 17:32:12
 */

/*
Note about freeing memory.
We need to free memory when we delete a node. But don't use "delete node"!
construct on an interview without discussing it with the interviewer.
A list node can be allocated in many different ways and we can use delete node;
only if we are sure that the nodes were allocated with new TreeNode(...).

We can create LinkList in stack as followings:
ListNode node1(2), node2(2), node3(2), node4(5), node5(7), node6(10);
node1.next = &node2; node2.next = &node3; ....
ListNode* p = Solution().deleteDuplicates(&node1);
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// Recursively
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(head==NULL || head->next==NULL){
            return head;
        }
        if(head->val == head->next->val){
            while(head->next && head->val==head->next->val){
                head = head->next;
            }
            return deleteDuplicates(head->next);
        }
        else{
            head->next = deleteDuplicates(head->next);
            return head;
        }
    }
};


// Ieratively
class Solution_2 {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode *pre_head = new ListNode(0);
        ListNode *cur = pre_head;
        while(head){
            if(head->next && head->val == head->next->val){
                while(head->next && head->val==head->next->val){
                    head=head->next;
                }
                head = head->next;
            }
            else{
                cur->next = head;
                head = head->next;
                cur = cur->next;
            }
            cur->next = NULL;
        }
        return pre_head->next;
    }
};

/*
[]
[1]
[1,2,2]
[3,3,3,3,3]
[1,1,1,2,3,4,4,4,4,5]
*/
